idna/src/deprecated.rs

```diff
@@ -0,0 +1,246 @@
+// Copyright 2013-2014 The rust-url developers.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Deprecated API for [*Unicode IDNA Compatibility Processing*
+//! (Unicode Technical Standard #46)](http://www.unicode.org/reports/tr46/)
+
+#![allow(deprecated)]
+
+use alloc::borrow::Cow;
+use alloc::string::String;
+
+use crate::uts46::*;
+use crate::Errors;
+
+/// Performs preprocessing equivalent to UTS 46 transitional processing
+/// if `transitional` is `true`. If `transitional` is `false`, merely
+/// lets the input pass through as-is.
+///
+/// The output of this function is to be passed to [`Uts46::process`].
+///
+/// Deprecated, since this functionality is deprecated in UTS 46 itself,
+/// and none of Firefox, Safari, or Chrome use transitional processing.
+#[deprecated]
+fn map_transitional(domain: &str, transitional: bool) -> Cow<'_, str> {
+    if transitional {
+        let mut chars = domain.chars();
+        loop {
+            let prev = chars.clone();
+            if let Some(c) = chars.next() {
+                match c {
+                    'ß' | 'ẞ' | 'ς' | '\u{200C}' | '\u{200D}' => {
+                        let mut s = String::with_capacity(domain.len());
+                        let tail = prev.as_str();
+                        let head = &domain[..domain.len() - tail.len()];
+                        s.push_str(head);
+                        for c in tail.chars() {
+                            match c {
+                                'ß' | 'ẞ' => {
+                                    s.push_str("ss");
+                                }
+                                'ς' => {
+                                    s.push('σ');
+                                }
+                                '\u{200C}' | '\u{200D}' => {}
+                                _ => {
+                                    s.push(c);
+                                }
+                            }
+                        }
+                        return Cow::Owned(s);
+                    }
+                    _ => {}
+                }
+            } else {
+                break;
+            }
+        }
+    }
+    Cow::Borrowed(domain)
+}
+
+/// Deprecated. Use the crate-top-level functions or [`Uts46`].
+#[derive(Default)]
+#[deprecated]
+pub struct Idna {
+    config: Config,
+}
+
+impl Idna {
+    pub fn new(config: Config) -> Self {
+        Self { config }
+    }
+
+    /// [UTS 46 ToASCII](http://www.unicode.org/reports/tr46/#ToASCII)
+    #[allow(clippy::wrong_self_convention)]
+    pub fn to_ascii(&mut self, domain: &str, out: &mut String) -> Result<(), Errors> {
```

Comment:
This is just the way the old API is. The purpose of deprecated.rs is to avoid breaking callers who just upgrade from 0.5.0 to 0.6.0 (assuming this becomes 0.6.0). I have tried not to fix the API in deprecated.rs.